### DISCUSSION TITLE
Concurrency concerns in asynchronous callbacks

### DISCUSSION THREAD

**Discussion by raffaeler** [eModbus/eModbus#335] [Q&A] [answered]
Hello, I am working on an ESP32 project and I see the worker manages the communication on a separate task.
I am wondering whether the client side callbacks (`handleData`, `handleError`) are called from a separate task.

If yes, I obviously have to protect all the data with a mutex to avoid inevitable race conditions.
If instead the call is queued from the main task, i can avoid this overhead.

Does anybody already knows the answer?
TIA

**Comment by Miq1** [comment]
I do, obviously 🤣 

The callbacks are executed in the worker task, so right, to protect against race conditions, you may need to use mutexes. But keep in mind the stack space of the worker task is limited to 4kB normally anyway (if you did not modify that), so the common solution is to park the data in a separate buffer and set a flag for the main task to inform it about it.

**Reply by raffaeler** [reply]
Thank you @Miq1 for confirming that.
I would strongly suggest to put a big warning in the docs ;-)

I already use tasks and mutexes and monitor their stacks. Thanks for the warning but it's not a big deal. Usually everything stays inside a class (mutex included) and the worker get a pointer to the class instance so that it can privately use data without overusing the stack.

**Reply by raffaeler** [reply]
BTW are the questions/answers sequence (ordering) to different servers guaranteed to be **preserved**?
I see the code waiting (eventually until timeout) right after the questions from the client.
Thanks again

**Reply by Miq1** [reply, marked as answer]
Yes (RTU). All requests are put on a single queue as they arrive and are processed strictly in order. As a consequence later requests will be held up when a server is slow or not responding.
It is different for TCP, because each connection to a server will have a separate client task.
